### ISSUE TITLE
de-flake and re-enable vanadium-mojo-v23proxy-integration-test 

### ISSUE THREAD

**Issue by caprita** [vanadium/issues#1226]
Disabled in the Jenkins UI, as per https://github.com/vanadium/build/issues/21
To re-enable, click "Enable" on https://veyron.corp.google.com/jenkins/job/vanadium-mojo-v23proxy-integration-test/

**Comment by bprosnitz** [comment]
@afandria assigned you to this because I believe you had already investigated this somewhat

**Comment by afandria** [comment]
Yeah, I'll take a look. If the flake is on Mojo Shell, we'll need to find a way around it with an internal timeout instead of hanging until the Jenkins test times out. If it's part of our code, we'll have to fix it ourselves.

I do think Mojo Shell sometimes has difficulty starting up binaries, so I'm going to investigate with that in mind.

**Comment by afandria** [comment]
FYI, re-enabling the test tentatively. I wasn't able to reproduce it locally after quite a few runs (v.io/c/20932 makes it runnable again).

**Comment by afandria** [comment]
The test still doesn't fail locally, but it fails quite often on Jenkins. It looks like endpoints are found, so it seems like Dart is the problem.

I've looked at the failures, and excluding some build and kubernetes issues, they fall into 2 categories
- Dart client fails to start (or hangs somewhere in the implementation) ([428](https://veyron.corp.google.com/jenkins/job/vanadium-mojo-v23proxy-integration-test/428/console), [441](https://veyron.corp.google.com/jenkins/job/vanadium-mojo-v23proxy-integration-test/441/console), [446](https://veyron.corp.google.com/jenkins/job/vanadium-mojo-v23proxy-integration-test/446/console), [451](https://veyron.corp.google.com/jenkins/job/vanadium-mojo-v23proxy-integration-test/451/console))
- Dart server implementation fails to start (or hangs somewhere in the implementation) ([427](https://veyron.corp.google.com/jenkins/job/vanadium-mojo-v23proxy-integration-test/427/console), [432](https://veyron.corp.google.com/jenkins/job/vanadium-mojo-v23proxy-integration-test/432/console), [450](https://veyron.corp.google.com/jenkins/job/vanadium-mojo-v23proxy-integration-test/450/console))

Note that the Dart server's service describer _does_ respond even when the Dart server implementation fails, so it's likely that the apps do start up and that the hang occurs elsewhere.

**Comment by afandria** [comment]
Oddly, the Dart server (on success) is cached twice by the network service, but only ever cached once if it fails.

A line like

```
16:20:57 [INFO:network_fetcher.cc(322)] Caching mojo app http://127.0.0.1:54711/dart-tests/end_to_end_test/lib/server.dart at /tmp/647129026/.mojo_url_response_disk_cache/cache/DF4C745F980C3A7F3BC1C19B0B9AC524BF78BFB9CA996A2E1DE09AB7DF306582/DF4C745F980C3A7F3BC1C19B0B9AC524BF78BFB9CA996A2E1DE09AB7DF306582
```

is seen twice on successful Dart server runs (with similar duplication for Go server runs). However, failures means that it only shows up once, presumably for the service describer.

In a similar vein, the caching of the client file occurs if the test passes and doesn't appear if there is a hang.

It's rather old, but it might be that multiprocess is causing these problems? (https://github.com/domokit/mojo/issues/508)

**Comment by afandria** [comment]
There's one other thing I can think of, and that's the test _order_. Recall how Mojo shells can't be reused unless there a different HOME directory is used.

It's possible that this problem is caused by the clients in the v23proxy test not setting their own HOME dir. Back to back runs may then lead to this hang occurring. (I thought sleeps may help, but perhaps Mojo Shell isn't even quitting.)

**Comment by afandria** [comment]
Unfortunately, this hasn't worked. :(

**Comment by jhahn21** [comment]
@afandria, didn't setting HOME to a temporary directory work?

**Comment by afandria** [comment]
It works to solve a different sort of hang (where two mojo shells can't start up because of a db lock), but it doesn't seem related to this failure because there were still hangs despite every shell getting its own unique HOME directory.
